### ISSUE TITLE
Possible issue with decoding emoji?

### ISSUE THREAD

**Issue by joelkuiper** [phronmophobic/llama.clj#6]
I'm trying to get the llma model to generate emoji. Which works

```
(base) build/bin [4329d1a] » ./main -m /array/Models/llama/llama-2-13b-chat/ggml-model-q4_1_m.bin -n 256 --repeat_penalty 1.0  -p "The unicode emoji for :smile:"
main: build = 950 (4329d1a)
main: seed  = 1692456952
ggml_init_cublas: found 1 CUDA devices:
  Device 0: NVIDIA GeForce RTX 2080 SUPER, compute capability 7.5
llama.cpp: loading model from /array/Models/llama/llama-2-13b-chat/ggml-model-q4_1_m.bin
llama_model_load_internal: format     = ggjt v3 (latest)
llama_model_load_internal: n_vocab    = 32000
llama_model_load_internal: n_ctx      = 512
llama_model_load_internal: n_embd     = 5120
llama_model_load_internal: n_mult     = 256
llama_model_load_internal: n_head     = 40
llama_model_load_internal: n_head_kv  = 40
llama_model_load_internal: n_layer    = 40
llama_model_load_internal: n_rot      = 128
llama_model_load_internal: n_gqa      = 1
llama_model_load_internal: rnorm_eps  = 5,0e-06
llama_model_load_internal: n_ff       = 13824
llama_model_load_internal: freq_base  = 10000,0
llama_model_load_internal: freq_scale = 1
llama_model_load_internal: ftype      = 3 (mostly Q4_1)
llama_model_load_internal: model size = 13B
llama_model_load_internal: ggml ctx size =    0,11 MB
llama_model_load_internal: using CUDA for GPU acceleration
llama_model_load_internal: mem required  = 8156,02 MB (+  400,00 MB per state)
llama_model_load_internal: offloading 0 repeating layers to GPU
llama_model_load_internal: offloaded 0/43 layers to GPU
llama_model_load_internal: total VRAM used: 360 MB
llama_new_context_with_model: kv self size  =  400,00 MB

system_info: n_threads = 24 / 48 | AVX = 1 | AVX2 = 1 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | VSX = 0 | 
sampling: repeat_last_n = 64, repeat_penalty = 1,000000, presence_penalty = 0,000000, frequency_penalty = 0,000000, top_k = 40, tfs_z = 1,000000, top_p = 0,950000, typical_p = 1,000000, temp = 0,800000, mirostat = 0, mirostat_lr = 0,100000, mirostat_ent = 5,000000
generate: n_ctx = 512, n_batch = 512, n_predict = 256, n_keep = 0


 The unicode emoji for :smile: is 😊.
The ASCII emoji for :smile: is 🙂. [end of text]

llama_print_timings:        load time =   647,97 ms
llama_print_timings:      sample time =     0,72 ms /    26 runs   (    0,03 ms per token, 35961,27 tokens per second)
llama_print_timings: prompt eval time =   429,78 ms /    11 tokens (   39,07 ms per token,    25,59 tokens per second)
llama_print_timings:        eval time =  4052,34 ms /    25 runs   (  162,09 ms per token,     6,17 tokens per second)
llama_print_timings:       total time =  4487,84 ms
```
The library was build with CUDA support and seems to be working fine.

However the same in Clojure yields: 

```
nREPL server started on port 40307 on host localhost - nrepl://localhost:40307
nREPL 0.9.0
Clojure 1.11.1
OpenJDK 64-Bit Server VM 19.0.2+7-Ubuntu-0ubuntu322.04
Interrupt: Control+C
Exit:      Control+D or (exit) or (quit)
(require '[com.phronemophobic.llama :as llama])
nil
(def context
  (llama/create-context
   "/array/Models/llama/llama-2-13b-chat/ggml-model-q4_1_m.bin"
   {:n-gpu-layers 1}))
ggml_init_cublas: found 1 CUDA devices:
  Device 0: NVIDIA GeForce RTX 2080 SUPER, compute capability 7.5
llama.cpp: loading model from /array/Models/llama/llama-2-13b-chat/ggml-model-q4_1_m.bin
llama_model_load_internal: format     = ggjt v3 (latest)
llama_model_load_internal: n_vocab    = 32000
llama_model_load_internal: n_ctx      = 512
llama_model_load_internal: n_embd     = 5120
llama_model_load_internal: n_mult     = 256
llama_model_load_internal: n_head     = 40
llama_model_load_internal: n_head_kv  = 40
llama_model_load_internal: n_layer    = 40
llama_model_load_internal: n_rot      = 128
llama_model_load_internal: n_gqa      = 1
llama_model_load_internal: rnorm_eps  = 5,0e-06
llama_model_load_internal: n_ff       = 13824
llama_model_load_internal: freq_base  = 10000,0
llama_model_load_internal: freq_scale = 1
llama_model_load_internal: ftype      = 3 (mostly Q4_1)
llama_model_load_internal: model size = 13B
llama_model_load_internal: ggml ctx size =    0,11 MB
llama_model_load_internal: using CUDA for GPU acceleration
llama_model_load_internal: mem required  = 7966,92 MB (+  400,00 MB per state)
llama_model_load_internal: allocating batch_size x (640 kB + n_ctx x 160 B) = 360 MB VRAM for the scratch buffer
llama_model_load_internal: offloading 1 repeating layers to GPU
llama_model_load_internal: offloaded 1/43 layers to GPU
llama_model_load_internal: total VRAM used: 550 MB
llama_new_context_with_model: kv self size  =  400,00 MB
#'user/context
(llama/generate-string context "The unicode emoji for :smile:")
", displayed on most modern devices.\n\n����"
user=> 
```
It does the � consistently for all emoji I checked. Which is weird given that it works in the cpp version! 

Only hiccup I found so far though, it's amazing stuff!

**Comment by phronmophobic** [comment]
Yes, that's a bug. Good catch!

The problem seems to be that each token is decoded as a string, but it seems that some tokens produce only partial unicode characters.

Here's a minimal reproduction:

```clojure
(def model-path "models/llama-2-7b-chat.ggmlv3.q4_0.bin")
(def ctx (create-context model-path {:n-gpu-layers 1}))

(defn token->pointer [tok]
  (.invoke 
   (.getFunction raw/libllama
                 "llama_token_to_str")
   Pointer (to-array [ctx (int tok)])))

(def tokens [243 162 155 135])
(def token-pointers (mapv token->pointer tokens))

(defn tokens->bytes [ps]
  (loop [bs []
         idx 0
         offset 0]
    (if-let [p (nth ps idx nil)]
      (let [b (.getByte p offset)]
        (if (zero? b)
          (recur bs (inc idx) 0)
          (recur (conj bs b) idx (inc offset))))
      bs)))

(clojure.string/join (eduction
                      (map #(raw/llama_token_to_str ctx %))
                      tokens))
;; "����"

(String. (byte-array (tokens->bytes token-pointers)) "utf-8") ;; "😄"
```

Fixing this for `llama/generate-string` shouldn't be too hard, but I'm not actually sure what to do for `llama/generate` since it's not easy to tell when you've collected enough to tokens to complete a unicode character. It's especially tricky for grapheme clusters like the emoji flag sequences. I'll have to think about it a bit.

Thanks for the report!

**Comment by phronmophobic** [comment]
I think I have a fix on the https://github.com/phronmophobic/llama.clj/tree/fix-decoding-tokens branch. I still want to test a few things before merging into `main`.

**Comment by phronmophobic** [comment]
Ok, should be fixed in version `0.6`.

**Comment by joelkuiper** [comment]
Hey @phronmophobic thank you so much! Sorry was afk for a bit, hence the late reply. Looks like it works now :smile: 

Many thanks for this great innovation, surely interesting stuff can be made with this now :upside_down_face:
